File: 1175. Prime Arrangements.cpp
class Solution {
public:
    const int M = 1e9 + 7;
    int numPrimeArrangements(int n) {
        vector<int> vis(n+1), p;
        for (int i=2; i<=n; ++i) {
            if (!vis[i]) p.push_back(i);
            for (int j=0; j<p.size() && p[j]*i<=n; ++j) {
                vis[p[j]*i] = true;
                if (i % p[j]==0) break;
            }
        }
        long long ans = 1;
        
        for (int i=2; i<=p.size(); ++i) 
            ans = ans * i % M;
        for (int i=2; i<=n-p.size(); ++i)
            ans = ans * i % M;
        return ans;
    }
};

File: 1175. Prime Arrangements.py
class Solution:
    def numPrimeArrangements(self, n: int) -> int:
        vis, p = [False] * (n+1), []
        for i in range(2, n+1):
            if not vis[i]: p.append(i)
            for j in range(len(p)):
                if p[j]*i <= n:
                    vis[p[j] * i] = True
                    if i%p[j]==0:
                        break
                else:
                    break
        ans = 1
        M = 1000000007
        for i in range(1, len(p)+1):
            ans = ans * i % M
        for i in range(1, n-len(p)+1):
            ans = ans * i % M
        return ans